internal/injector/parse/parser.go

```diff
@@ -0,0 +1,204 @@
+
+import (
+	"errors"
+	"fmt"
+	"go/ast"
+	goparser "go/parser"
+	"go/token"
+	"io"
+	"os"
+	"slices"
+	"sync"
+	"sync/atomic"
+
+	"github.com/DataDog/orchestrion/internal/injector/aspect"
+	"github.com/DataDog/orchestrion/internal/injector/aspect/may"
+)
+
+const maxBytesEagerness = 1 << 19 // 512 KiB
+
+type rawFile struct {
+	name       string
+	mappedName string
+	content    []byte
+}
+
+type File struct {
+	Name    string
+	AstFile *ast.File
+	Aspects []*aspect.Aspect
+}
+
+type Parser struct {
+	fset *token.FileSet // thread-safe data structure
+
+	// rawFiles is an intermediary data structure to store the raw content of the files before parsing them.
+	rawFiles []rawFile
+
+	// filesBytesCount is the sum of the bytes of all files parsed so far.
+	filesBytesCount atomic.Uint64
+
+	// eagerness is a flag that is set to true if at least one file has been parsed and requires all files to be parsed.
+	eagerness atomic.Bool
+
+	// parsedFiles is what is returned by ParseFiles.
+	parsedFiles []File
+
+	wg sync.WaitGroup
+
+	errs   []error
+	errsMu sync.Mutex
+}
+
+func NewParser(fset *token.FileSet, nbFiles int) *Parser {
+	p := &Parser{
+		fset:        fset,
+		rawFiles:    make([]rawFile, nbFiles),
+		parsedFiles: make([]File, nbFiles),
+	}
+
+	p.eagerness.Store(true)
+
+	return p
+}
+
+// ParseFiles return either zero files or all files parsed with their respective aspects
+func (p *Parser) ParseFiles(files []string, aspects []*aspect.Aspect) ([]File, error) {
+	p.wg.Add(len(files))
+
+	for idx, file := range files {
+		fileAspects := make([]*aspect.Aspect, len(aspects))
+		copy(fileAspects, aspects)
+
+		go func(idx int, file string) {
+			defer p.wg.Done()
+			var err error
+			p.rawFiles[idx], err = readFile(file)
+			if err != nil {
+				p.addError(fmt.Errorf("reading %q: %w", file, err))
+				return
+			}
+
+			p.filesBytesCount.Add(uint64(len(p.rawFiles[idx].content)))
+			if p.aspectMayMatch() {
+				fileAspects = p.fileFilterAspects(fileAspects, p.rawFiles[idx])
+				if len(fileAspects) == 0 {
+					// No aspects can match on this file, no need to fill up the File.AstFile field.
+					p.parsedFiles[idx] = File{Name: file}
+					return
+				}
+			}
+
+			p.eagerness.Store(false)
+			p.parsedFiles[idx] = p.parseFile(p.rawFiles[idx], fileAspects)
+		}(idx, file)
+	}
+
+	p.wg.Wait()
+
+	// No aspects can match on this package, return nothing
+	if !p.aspectMayMatch() {
+		return nil, nil
+	}
+
+	if len(p.errs) > 0 {
+		return nil, errors.Join(p.errs...)
+	}
+
+	// If we arrived here, this means we need to parse all files anyway because the type-checking pass will need them.
+	p.parseMissingFiles()
+	p.wg.Wait()
+
+	return p.parsedFiles, errors.Join(p.errs...)
+}
+
+// aspectMayMatch returns true if the parser should parse all files because at least one file requires it.
+func (p *Parser) aspectMayMatch() bool {
+	return !p.eagerness.Load() || p.filesBytesCount.Load() > maxBytesEagerness
+}
+
+func (p *Parser) addError(err error) {
```

Comment:
1. All usage of this actually pass the result of an `fmt.Errorf` call, so you could have the signature here be:
   ```suggestion
   func (p *Parser) addError(format string, args... any) {
   ```
2. If you use `errgroup.Group` as suggested previously, I think this becomes redundant.